tests/server/query_params/test_filter.py

```diff
@@ -378,8 +378,109 @@ def test_not_or_and_precedence(check_response):
     expected_ids = ["mpf_1", "mpf_200"]
     check_response(request, expected_ids)
 
-    request = '/structures?filter=elements HAS "Ac" AND nelements>1 AND nsites=1'
-    expected_ids = []
+    request = '/structures?filter=elements HAS "Ac" AND nelements>1 AND nsites=3'
+    expected_ids = ["mpf_23"]
+    check_response(request, expected_ids)
+
+
+def test_behaviour_not(check_response, client, check_error_response):
+    request = '/structures?filter=NOT(elements HAS "Ag" AND nelements>1 )'
+    expected_ids = [
+        "mpf_1",
+        "mpf_23",
+        "mpf_30",
+        "mpf_110",
+        "mpf_3803",
+        "mpf_3819",
+        "mpf_200",
+    ]
+    check_response(request, expected_ids)
+
+    request = (
+        '/structures?filter=NOT(references.id HAS ALL "dummy/2019", "dijkstra1968")'
+    )
+    expected_ids = ["mpf_1", "mpf_2", "mpf_3", "mpf_3819"]
+    if CONFIG.database_backend == SupportedBackend.ELASTIC:
+        check_error_response(
+            request,
+            expected_status=501,
+            expected_title="NotImplementedError",
+            expected_detail="Unable to filter on relationships with type 'references'",
+        )
+        pytest.xfail("Elasticsearch backend does not support relationship filtering.")
+    check_response(request, expected_ids)
+
+    request = '/structures?filter=NOT(elements HAS ALL "Ag", "N")'
+    expected_ids = [
+        "mpf_1",
+        "mpf_2",
+        "mpf_3",
+        "mpf_23",
+        "mpf_30",
+        "mpf_110",
+        "mpf_200",
+        "mpf_220",
+        "mpf_446",
+        "mpf_3803",
+        "mpf_3819",
+    ]
+    check_response(request, expected_ids)
+
+    request = '/structures?filter=NOT(elements HAS "Ac" AND nelements>1 AND nsites=1)'
+    expected_ids = [
+        structure["id"] for structure in client.get("/structures").json()["data"]
+    ]
+    check_response(request, expected_ids)
+
+    request = '/structures?filter=NOT(elements HAS "Ac" OR nelements>1 OR nsites>1)'
+    expected_ids = ["mpf_200"]
+    check_response(request, expected_ids)
+
+    request = "/structures?filter=NOT(nsites<4 OR nsites>20 AND NOT(nelements >5))"
```

Comment:
(re: comment below, mongomock seems to have no problem with the double negation at the end here?)

Reply:
This query is simplified. When the transformer encounters NOT(nelements > 5) It gets transformed into {"nelements":{"$lte":5}} So a double NOT, e.g. {$not:{$not: ...}}, does not occur in this case. It only happens when there is a not statement I cannot simplify, for example NOT(NOT(A CONTAINS "B")).  There is no opposite of the CONTAINS operator, so I have to leave the $not in front of it. Of course, it should be possible to remove these double not statements in our transformer. I, however, already made a patch for mongomock, so I hoped this would not be neccesary.